### PR TITLE
Enterprise Testing System

### PR DESCRIPTION
**Pull request by davidsh**
This is the first of several PRs that add Enterprise Scenarios Testing capability to
the repo. This PR focusses on Linux which allows for docker containers to be used
in an enterprise network configuration.

I focussed on 2 workflows: 1) The 'dev' workflow, and 2) The PR/CI workflow. The dev
workflow works well since it's using containers in a docker-compose environment along
with volume mounting your current dev's repo enlistment. The PR/CI workflow gives us
an Azure DevOps pipeline to automate verification.

I still need to work with the infra team to add a real pipeline that will run. I can't
do that until this is merged. In the meantime, I have my own DevOps pipeline that verified this PR.

See: https://dev.azure.com/systemnetncl/Enterprise%20Testing/_build/results?buildId=141

I will be linking a follow-up GitHub issue describing the roadmap for building on this system
including adding Windows environments, NTLM protocol, proxies, and other libraries such as
System.Net.Mail and System.Data.SqlClient. Those libraries also use Negotiate/Kerberos/NTLM
enterprise-oriented protocols.

Contributes to:
https://github.com/dotnet/corefx/issues/41652
https://github.com/dotnet/corefx/issues/41489
https://github.com/dotnet/corefx/issues/36896
https://github.com/dotnet/corefx/issues/30150
https://github.com/dotnet/corefx/issues/24707
https://github.com/dotnet/corefx/issues/10041
https://github.com/dotnet/corefx/issues/6606
https://github.com/dotnet/corefx/issues/6161

### CODE REVIEW

**Comment by davidsh** [comment]
cc: @danmosemsft @karelz

**Comment by davidsh** [comment]
@Tratcher

**Comment by davidsh** [comment]
fyi @mconnew

**Review by scalablecory — APPROVED** [review]
I'm not familiar enough with Docker to approve those scripts, but the C# looks fine to me.

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/apacheweb/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+# Prevents dialog prompting when installing packages
+ARG DEBIAN_FRONTEND=noninteractive
+
+# Install Kerberos client, apache Negotiate auth plugin, and diagnostics
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends libapache2-mod-auth-kerb procps krb5-user iputils-ping dnsutils nano
+
+# Copy apache2 kerb module to the right place since the apt-get install puts it in the wrong place for this docker image
+RUN cp /usr/lib/apache2/modules/mod_auth_kerb.so /usr/local/apache2/modules
```
Comment (wfurt) [inline src/libraries/Common/tests/System/Net/EnterpriseTests/setup/apacheweb/Dockerfile]: `ln -s src dst` can work as well. and that preserves linkage to the source.

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/linuxclient/run.sh [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+cp /SHARED/linuxclient.keytab /etc/krb5.keytab
+
+# Keep the container running
+tail -f /dev/null
```
Comment (wfurt) [inline src/libraries/Common/tests/System/Net/EnterpriseTests/setup/linuxclient/run.sh]: this looks strange. often you can simply run bash so when executed interactively you get shell.

**Review by wfurt — APPROVED** [review]
LGTM.

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/linuxclient/run.sh (hunk 2 above)
Comment (davidsh) [reply src/libraries/Common/tests/System/Net/EnterpriseTests/setup/linuxclient/run.sh]: Yes, it is strange. :) But I got the tip from looking at different containers everywhere.

I use it because 'docker-compose up -d' keeps the cluster of machines running in the background. And specifying only 'bash' as the EXEC/CMD of the container doesn't work for detached containers. It would cause the container to exit immediately.

The workflow is that once the containers are up and running in a detached mode, then one can access the containers by doing "docker exec -it linuxclient bash" (for example). That gives you a bash session where you are equivalently "logged on" to the container much like SSH. And then doing 'exit' from that drops you back to your host machine, but keeps the container running (as designed).

When you are done with all the containers, doing a 'docker-compose down' shuts them all down and removes the private network.

I documented this workflow in the README.md.

**Review by Tratcher — APPROVED** [review]
Can you summarize the network architecture so far? Is this all set up on a single container? Or do you have a separate container for the KDC?
How much more do you think you'd need to do to get a test like this one working?  https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Security/tests/FunctionalTests/NegotiateStreamKerberosTest.cs#L169

**Comment by davidsh** [comment]
@Tratcher 

>Is this all set up on a single container? Or do you have a separate container for the KDC?

Currently this PR sets up a docker-compose network of 3 containers: kdc, apachewebserver, linuxclient. That is sufficient to run some tests. Strictly speaking the kdc could be merged together with the webserver but I kept things separate. But the workflow is that the tests run on the "client" machine. This is explained in the README.md file.

>How much more do you think you'd need to do to get a test like this one working? https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Security/tests/FunctionalTests/NegotiateStreamKerberosTest.cs#L169

For that particular test (which tests the server-side of NegotiateStream), I am already testing that server-side NegotiateStream scenario exactly as part of the new NegotiateStreamLoopbackTest.cs (this PR). It's just that it is doing it using the VirtualNetwork instead of a real TCP socket. I plan to add more NegotiateStream tests in a later PR to demonstrate more end-to-end testing.

I also have implemented another linux docker container that has ASP.NET Core running as a web server. And both Negotiate and NTLM (via Negotiate's Kerberos to NTLM fallback) is working on that now. So, I will add more tests to the HttpClient tests that verify that scenario (client talking to authenticating ASP.NET Core web server). I will add that docker container as part of this "enterprise network" of docker containers.

>Can you summarize the network architecture so far?

I currently have the README.md to explain how the dev workflow and CI workflow works. But I plan to write up soon a larger document to explain the architectures.

This PR establishes the current multiple docker container architecture which works well for Linux only. But Windows requires real VMs for Active Directory scenario. For that, we are now hosting a multi-VM cluster of machines running Active Directory. This runs on a dedicated and funded Azure subscription for network testing. We will install Azure DevOps agents on the client machine on these VMs which will then run these tests automatically in CI for Windows and Windows/Linux scenarios. That is not fully implemented yet.

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+## Networking Components of Enterprise Scenarios
+Enterprise scenarios typically see the following kinds of components/protocols/security:
+* Although possibly connected to the Internet, most of the networking topology is internal facing. There is use of some internal “directory” service for authentication of connected computers and users. On Windows, this can include Windows Active Directory domains with computers being domain-controllers, domain-joined, or standalone computers. On Linux, this includes Kerberos realms using KDCs and participating member computers. With .NET Core being cross-platform, this now includes connections with multiple domains and realms with various cross trust between them.
+* Authentication protocols such as NTLM, Kerberos and Negotiate. These are used more than Basic and Digest. Negotiate/Kerberos requires both client and server computers to be “joined” to a common trusted directory service.
+* TLS/SSL extensively used.
```
[the remark this answers is not in the document]
Comment (davidsh) [reply src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md]: Thanks for the feedback. This document was written to help explain the overall concept of 'enterprise scenarios' to help developers new to the repo and these scenarios. I'll try to revise the text per your feedback.

**Comment by Tratcher** [comment]
Cool, thanks. I have some cross-machine tests using HttpClient and Kestrel that still need an environment to run in. You may be able to poach from there.
https://github.com/aspnet/AspNetCore/blob/master/src/Security/Authentication/Negotiate/test/Negotiate.FunctionalTest/CrossMachineReadMe.md

**Comment by davidsh** [comment]
>Cool, thanks. I have some cross-machine tests using HttpClient and Kestrel that still need an environment to run in. You may be able to poach from there.

Thanks for permission for "poaching" them. I was planning that. :)

File: eng/pipelines/libraries/enterprise/linux.yml [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    - src/installer/*
+    - src/coreclr/*
+    - eng/pipelines/coreclr/*
+    - eng/pipelines/installer/*
+    - eng/pipelines/common/*
+
+pr:
+  branches:
+    include:
+    - master
```
Comment (eiriktsarpalis) [inline eng/pipelines/libraries/enterprise/linux.yml]: Do we want these tests to trigger automatically on every PR ?

File: eng/pipelines/libraries/enterprise/linux.yml [hunk 5]
```diff
@@ -0,0 +1,97 @@
+trigger:
+  batch: true
+  branches:
+    include:
+    - master
```
Comment (eiriktsarpalis) [inline eng/pipelines/libraries/enterprise/linux.yml]: Do we want CI to trigger on every push to master?

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+
+## Running the tests
+These tests need to be run in the dedicated Enterprise Test environment. This environment can be created on a local dev machine as long as Docker in installed. The enterprise test environment is a collection of Linux docker containers (client and servers) connected on a common docker network.
+
+Set the DOTNET_RUNTIME_REPO_ROOT environment variable to the path on your dev machine (the host machine) where the repo is installed:
+
+```
+# Windows cmd.exe shell example
+set DOTNET_RUNTIME_REPO_ROOT=s:\GitHub\runtime
```
Comment (eiriktsarpalis) [inline src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md]: Why is this variable required? Given that all the scripts are already in that repo, it's pretty straightforward to infer this information.

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/kdc/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,27 @@
+FROM mcr.microsoft.com/dotnet/core/sdk:3.0.100-bionic
```
Comment (eiriktsarpalis) [inline src/libraries/Common/tests/System/Net/EnterpriseTests/setup/kdc/Dockerfile]: Is dotnet sdk required for this container? Looks like it could work with a vanilla ubuntu image.

File: eng/pipelines/libraries/enterprise/linux.yml [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+- bash: |
+    cd $(enterpriseTestsSetup)
+    docker-compose up -d
+  displayName: Start test network and machines
+  env:
+    DOTNET_RUNTIME_REPO_ROOT: $(Build.SourcesDirectory)
+
+- bash: |
+    docker exec linuxclient bash /setup/test-webserver.sh
```
Comment (eiriktsarpalis) [inline eng/pipelines/libraries/enterprise/linux.yml]: Your use of docker exec for controlling the container is interesting, and I can probably see why you went for that approach. While it does offer flexibility and convenience, I'm wondering whether it can cause problems, particularly given the fact that it's just mounting the source files from the host filesystem. This has caused problems for me in the past, for example when building a dirty windows repo from WSL (build artifacts generated by windows confuse the linux sdk).

In the interest of reproducibility, I would recommend using a docker build which copies and builds the artifacts into a new image. The tests can then be run by calling `docker run` on the new image. Note that you can still override the `CMD` command with `docker run`, as well as issue new commands using `docker exec` while it is running. This has a few benefits, namely you can deploy the client to remote machines that do not necessarily have the source code checked out.

That being said, there is a trade-off, namely the copying of source files will certainly increase the build time. That should not be a concern from a pipelines perspective. I can see you might be motivated by setting up an interactive dev experience, but I think the same can be achieved by using, say, VsCode's RemoteSSH against a dedicated linux box.

**Comment by eiriktsarpalis** [comment]
> I still need to work with the infra team to add a real pipeline that will run. I can't
do that until this is merged.

You can debug your pipeline files by pushing your branch to the dotnet repo. It's frowned upon in general, but tolerated as long as it addresses infra concerns. It will certainly save you from having to post new PRs if a pipeline issue is discovered.

**Comment by eiriktsarpalis** [comment]
@davidsh I like your choice of folder structure. I'm wondering if `src/libraries/System/Net/Stress` might be an appropriate home for stress testing as well.

**Review by eiriktsarpalis — APPROVED** [review]
A few comments, but generally a very good addition. Would recommend testing the pipelines before merging in though.

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md (hunk 6 above)
Comment (davidsh) [reply src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md]: Many people won't look at the scripts. I wanted a set of instructions that was simple to follow. And setting this environment variable is required for the 'dev' workflow in order for the client container to build and run the tests.

File: eng/pipelines/libraries/enterprise/linux.yml (hunk 8 above)
Comment (davidsh) [reply eng/pipelines/libraries/enterprise/linux.yml]: >I would recommend using a docker build which copies and builds the artifacts into a new image.

I considered many approaches to this including the one you mention here which is used by the stress testing app. But I prioritized the 'dev' experience and wanted the experience to be as similar as possible to running the other 'FunctionalTests' we have in the repo. I also wanted it to be fast so that developers would be inclined to use it. And this technique allows for tight 'inner loop' development. You can quickly make a change to a few source files in a library and just build that library (inside the container). And then run the tests again.

>I'm wondering whether it can cause problems, particularly given the fact that it's just mounting the source files from the host filesystem. This has caused problems for me in the past, for example when building a dirty windows repo from WSL (build artifacts generated by windows confuse the linux sdk).

This is a good point. But I think it can be addressed with documentation here to give developers some tips when running the tests on their local dev box. And the technique being used here to build the repo and run the tests by mounting from the host works perfectly fine in CI on the Azure DevOps agent's machines. Those are always using a clean VM.

**Comment by davidsh** [comment]
>You can debug your pipeline files by pushing your branch to the dotnet repo. It's frowned upon in general, but tolerated as long as it addresses infra concerns. It will certainly save you from having to post new PRs if a pipeline issue is discovered.

Thanks for the reminder about branches on the main repo. I chose not to do that for now. But I will do that for some subsequent PRs as I tune the *.yml code for the pipeline. Right now, there is no actual dotnet/runtime pipeline for this *.yml code to run against.

File: eng/pipelines/libraries/enterprise/linux.yml (hunk 4 above)
Comment (davidsh) [reply eng/pipelines/libraries/enterprise/linux.yml]: My goal is 'yes'. I would like these tests to run for every PR in the same way we run innerloop tests in general. I consider these new 'enterprise scenario' tests to be innerloop.

File: eng/pipelines/libraries/enterprise/linux.yml (hunk 5 above)
Comment (davidsh) [reply eng/pipelines/libraries/enterprise/linux.yml]: I might adjust that after discussions with @safern and other infra folks. The primary goal is for these tests to be run for validating PRs. So, I'm not sure what value the pipeline trigger would be for other pushes to master. Besides PRs, what other kinds of pushes to master are there now?

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/kdc/Dockerfile (hunk 7 above)
Comment (davidsh) [reply src/libraries/Common/tests/System/Net/EnterpriseTests/setup/kdc/Dockerfile]: >Is dotnet sdk required for this container?

It isn't required. It was when I was first developing this solution and I was using the prior technique of having the tests build similar to the stress app (i.e. with public 'dotnet' tools). But I moved away from that approach.

I will  adjust this container to be based on regular ubuntu image in a follow-up PR as you suggest.

**Comment by davidsh** [comment]
>@davidsh I like your choice of folder structure. I'm wondering if src/libraries/System/Net/Stress might be an appropriate home for stress testing as well.

I'm not sure I understand what you are asking.  I don't have anything directly off the repo root of 'src/libraries' paths.

I have the actual tests in a folder called 'EnterpriseTests' that is parallel to 'FunctionalTests' for a particular library. And the setup code for the docker-compose machines/network is in 'src/libraries/Common/tests/System/Net/EnterpriseTests' folder.

Can you clarify what new paths you are proposing for the 'stress' testing?

**Comment by eiriktsarpalis** [comment]
> I have the actual tests in a folder called 'EnterpriseTests' that is parallel to 'FunctionalTests' for a particular library. And the setup code for the docker-compose machines/network is in 'src/libraries/Common/tests/System/Net/EnterpriseTests' folder.

I mistyped, I meant to say the `src/libraries/Common/tests/System/Net/EnterpriseTests` folder.

**Review by stephentoub — APPROVED** [review]

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md (hunk 6 above)
Comment (eiriktsarpalis) [reply src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md]: Not sure I follow. We have a lot of scripts that do this already, for example https://github.com/dotnet/runtime/blob/875a54a112999004418ca369dfb0c31a7612c3cd/eng/build.sh#L5-L13

**Comment by davidsh** [comment]
@eiriktsarpalis 

>I mistyped, I meant to say the src/libraries/Common/tests/System/Net/EnterpriseTests folder.

If you need common configuration/setup for stress tests, then putting them in new, parallel, folder like 'src/libraries/Common/tests/System/Net/StressTests' would be fine.

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+* Complex DNS architectures using various A and CNAME (alias) records.
+* Use of the NegotiateStream class. This is typically seen as part of using WCF client/server architecture.
+* Impersonation/Delegation of credentials. This occurs frequently in middle-tier scenarios that involve a client computer talking with a middle-tier computer (such as a web server) which makes an outbound call to another computer such as a database server. In this case, the credentials of the client computer are delegated across the middle-tier computer so that the outbound call to the database server is made in the context of the client’s credentials.
+
+
+## Running the tests
+These tests need to be run in the dedicated Enterprise Test environment. This environment can be created on a local dev machine as long as Docker in installed. The enterprise test environment is a collection of Linux docker containers (client and servers) connected on a common docker network.
+
+Set the DOTNET_RUNTIME_REPO_ROOT environment variable to the path on your dev machine (the host machine) where the repo is installed:
```
Comment (mconnew) [inline src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md]: Instead of making this explicit, could you have a simple script which runs the docker-compose command and at the start of the script sets this environment variable based on a path relative to the scripts location?

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+```
+
+Now you can run the enterprise tests. Currently, there are tests for System.Net.Http and System.Net.Security. You can run them in the same way you already run tests in the repo.
+
+
+(System.Net.Http example shown)
+
+```bash
+cd /repo/src/libraries/System.Net.Http/tests/EnterpriseTests
+/repo/.dotnet/dotnet msbuild /t:rebuilandtest
```
Comment (mconnew) [inline src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md]: typo, rebuil**d**andtest

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/apacheweb/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,25 @@
+FROM httpd:2.4
```
Comment (mconnew) [inline src/libraries/Common/tests/System/Net/EnterpriseTests/setup/apacheweb/Dockerfile]: Could you create a docker image with all this already done? Azure container registry has the ability to automatically rebuild the container based on triggers such as the base image being updated. See https://docs.microsoft.com/en-us/azure/container-registry/container-registry-tasks-overview

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md (hunk 10 above)
Comment (davidsh) [reply src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md]: Thanks.  Fixed.

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/apacheweb/Dockerfile (hunk 11 above)
Comment (davidsh) [reply src/libraries/Common/tests/System/Net/EnterpriseTests/setup/apacheweb/Dockerfile]: We're looking into caching some of these images to reduce image build time, especially for the 'linuxclient' container. However, once built on the local dev machine, it is cached by the docker system on the host machine.

It's always a trade-off with pre-building and storing these images since it creates another place to maintain infrastructure. We do have our existing Azure Container Registry, mcr.microsoft.com/dotnet-buildtools/prereqs, which might use for this purpose. This will be a workitem for a future PR after discussions with our infra group.

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md (hunk 9 above)
Comment (davidsh) [reply src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md]: It would be nice if 'docker-compose' had a built-in feature for the compose file that would give it a self-referencing path. Then we could easily do as you suggest.

For now, to achieve what you suggest, I'll need to create a wrapper script (bash and PowerShell) that would wrap the 'docker-compose' commands ('build', 'up -d', 'down') and would also set the environment variable. I'll look into that.

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/apacheweb/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+# Install Kerberos client, apache Negotiate auth plugin, and diagnostics
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends libapache2-mod-auth-kerb procps krb5-user iputils-ping dnsutils nano
+
+# Copy apache2 kerb module to the right place since the apt-get install puts it in the wrong place for this docker image
+RUN cp /usr/lib/apache2/modules/mod_auth_kerb.so /usr/local/apache2/modules
+
+# Modify httpd.conf to add Negotiate auth as required
+RUN echo "LoadModule auth_kerb_module modules/mod_auth_kerb.so" >> /usr/local/apache2/conf/httpd.conf && \
+    sed -i 's/Require all granted/AuthType Kerberos\nAuthName "Kerberos Login"\nKrbAuthRealm LINUX\.CONTOSO\.COM\nKrb5Keytab \/etc\/krb5\.keytab\nKrbMethodK5Passwd off\nRequire valid-user/' /usr/local/apache2/conf/httpd.conf
```
Comment (lpereira) [inline src/libraries/Common/tests/System/Net/EnterpriseTests/setup/apacheweb/Dockerfile]: TIL contoso.com :)

File: eng/pipelines/libraries/enterprise/linux.yml (hunk 5 above)
Comment (davidsh) [reply eng/pipelines/libraries/enterprise/linux.yml]: For this PR, I will revise the .yml file for the pipeline to disable running on pushes to branches. We'll just let it run on PRs.

The pipeline can also be manually run of course.

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/apacheweb/Dockerfile (hunk 12 above)
Comment (mconnew) [reply src/libraries/Common/tests/System/Net/EnterpriseTests/setup/apacheweb/Dockerfile]: @lpereira, you might enjoy [this wikipedia article](https://en.wikipedia.org/wiki/List_of_fictional_Microsoft_companies#Companies) :grin:

File: eng/pipelines/libraries/enterprise/linux.yml (hunk 4 above)
Comment (davidsh) [reply eng/pipelines/libraries/enterprise/linux.yml]: Since this is a new pipeline, I'm going to initially restrict the 'auto-run on PR' feature to just changes in certain high-priority network-related filepaths in the repo. As we gain experience from this feature, we'll adjust over time.

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/apacheweb/Dockerfile (hunk 12 above)
Comment (davidsh) [reply src/libraries/Common/tests/System/Net/EnterpriseTests/setup/apacheweb/Dockerfile]: I explicitly chose 'CONTOSO.COM' because it has no branding/copyright/ownership concerns when used in this context for testing.

**Review by safern — APPROVED** [review]
Note that there is a lot of changes going on with our CI and build logic so initially once: https://github.com/dotnet/runtime/pull/494 is merged you will need to change the way you build stuff in order to include a locally built coreclr, so this build will be broken since we're removing the ability to build from a coreclr package.

There are other changes that we're doing in order to achieve that, i.e: https://github.com/dotnet/runtime/pull/530, https://github.com/dotnet/runtime/pull/526 

Of course we can sync up and work together to do the required updates in order to achieve that.

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md (hunk 9 above)
Comment (davidsh) [reply src/libraries/Common/tests/System/Net/EnterpriseTests/setup/README.md]: Looks like we have some existing examples of helper bash and PowerShell scripts that automatically can determine the repo root (via git commands) and set an environment variable.

* https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/tests/StressTests/HttpStress/load-corefx-testhost.ps1
* https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/tests/StressTests/HttpStress/load-corefx-testhost.sh

I'll add scripts similar to that in a follow-up PR.

File: src/libraries/Common/tests/System/Net/EnterpriseTests/setup/apacheweb/Dockerfile (hunk 1 above)
Comment (davidsh) [reply src/libraries/Common/tests/System/Net/EnterpriseTests/setup/apacheweb/Dockerfile]: Thanks. Fixed.

File: eng/pipelines/libraries/enterprise/linux.yml [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+# To reduce load on the pipeline, enable it only for PRs that affect critical networking code
+pr:
+  branches:
+    include:
+    - master
+    - release/*.*
+
+  paths:
+    include:
+      - src/libraries/Common/src/System/Net/*
```
Comment (safern) [inline eng/pipelines/libraries/enterprise/linux.yml]: You also need to include `eng/pipelines/libraries/enterprise/*` to make sure when this file is changed you get the necessary coverage.

Comment (davidsh) [reply eng/pipelines/libraries/enterprise/linux.yml]: Thx. I'll add that in my follow-up PR.

Comment (davidsh) [reply eng/pipelines/libraries/enterprise/linux.yml]: Fixed in #843
